### ISSUE TITLE
[SPARK] Null pointer exception using learning rate schedule on Spark

### ISSUE THREAD

**Issue by sethah** [deeplearning4j/deeplearning4j#3420]
#### Issue Description
Ran with the following code:

````scala
val confBuilder = new NeuralNetConfiguration.Builder()
        .seed(params.seed)
        .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
        .iterations(1)
        .activation(Activation.SOFTMAX)
        .weightInit(WeightInit.XAVIER)
        .learningRate(params.learningRate)
        .learningRateDecayPolicy(LearningRatePolicy.Schedule)  
        .learningRateSchedule(schedule)  
        .updater(params.updater)
        .momentum(params.momentum)
        .regularization(true)
        .l2(params.regularization)
        .graphBuilder()
        .addLayer("predictions",
          new OutputLayer.Builder(LossFunction.NEGATIVELOGLIKELIHOOD)
            .activation(Activation.SOFTMAX)
            .nIn(4096)
            .nOut(257)
            .build(),
          "in")
        .setOutputs("predictions")
        .addInputs("in")
        .backprop(true)
        .pretrain(false)
        .build()
      val graph = new ComputationGraph(conf)
      val tm = new ParameterAveragingTrainingMaster.Builder(1)
        .averagingFrequency(params.averagingFrequency)
        .workerPrefetchNumBatches(2)
        .batchSizePerWorker(params.batchSizePerWorker)
        .rddTrainingApproach(RDDTrainingApproach.Export)
        .build()
     val model = new SparkComputationGraph(sc, graph, tm)
(0 until params.numEpochs).foreach { i =>
  model.fit(rdd)
}
````

Spark submit args

````
spark2-submit \
--master yarn \
--deploy-mode client \
--conf spark.driver.extraJavaOptions="-XX:+UseG1GC -XX:MaxGCPauseMillis=200 -Dorg.bytedeco.javacpp.maxretries=100 -Dorg.bytedeco.javacpp.maxbytes=13000000000" \
--conf spark.executor.extraJavaOptions="-XX:+UseG1GC -XX:MaxGCPauseMillis=200 -Dorg.bytedeco.javacpp.maxretries=100 -Dorg.bytedeco.javacpp.maxbytes=13000000000" \
--conf spark.locality.wait=0 \
--conf spark.driver.maxResultSize=10g \
--conf spark.yarn.executor.memoryOverhead=15g \
--conf spark.yarn.driver.memoryOverhead=15g \
--conf spark.executor.cores=8 \
--conf spark.kryo.registrator=org.nd4j.Nd4jRegistrator \
--conf spark.serializer=org.apache.spark.serializer.KryoSerializer \
--conf spark.kryoserializer.buffer.max=2047m \
--executor-memory 10g \
--driver-memory 10g \
--num-executors=5 \
````

Error:

````
Caused by: com.esotericsoftware.kryo.KryoException: java.lang.NullPointerException
Serialization trace:
learningRateSchedule (org.deeplearning4j.nn.conf.layers.BatchNormalization)
layer (org.deeplearning4j.nn.conf.NeuralNetConfiguration)
layerConf (org.deeplearning4j.nn.conf.graph.LayerVertex)
vertices (org.deeplearning4j.nn.conf.ComputationGraphConfiguration)
graphConfiguration (org.deeplearning4j.spark.api.worker.NetBroadcastTuple)
	at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:144)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:551)
````

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version: 0.8.0
* platform information (OS, etc): CentOS Linux 7 (Core)
* Spark 2.1.0
* YARN cluster manager, on top of CDH

**Comment by sethah** [comment]
cc @huitseeker

**Comment by jasstionzyf** [comment]
i also have such issues.

run with following code:

```
int batchSizePerWorker=Integer.parseInt(settings.getProperty("batchSizePerWorker","16"));
        int numEpochs =Integer.parseInt(settings.getProperty("numEpochs","1"));

        //Load the data into memory then parallelize

        //This isn't a good approach in general - but is simple to use for this example
        DataSetIterator iterTrain = new MnistDataSetIterator(batchSizePerWorker, true, 12345);
        DataSetIterator iterTest = new MnistDataSetIterator(batchSizePerWorker, true, 12345);
        List<DataSet> trainDataList = new ArrayList<>();
        List<DataSet> testDataList = new ArrayList<>();
        while (iterTrain.hasNext()) {
            trainDataList.add(iterTrain.next());
        }
        while (iterTest.hasNext()) {
            testDataList.add(iterTest.next());
        }



        JavaRDD<DataSet> trainData = sc.parallelize(trainDataList);
        JavaRDD<DataSet> testData = sc.parallelize(testDataList);


        //----------------------------------
        //Create network configuration and conduct network training
        MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
                .seed(12345)
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT).iterations(1)
                .activation(Activation.LEAKYRELU)
                .weightInit(WeightInit.XAVIER)
                .learningRate(0.02)
                .updater(Updater.NESTEROVS).momentum(0.9)
                .regularization(true).l2(1e-4)
                .list()
                .layer(0, new DenseLayer.Builder().nIn(28 * 28).nOut(500).build())
                .layer(1, new DenseLayer.Builder().nIn(500).nOut(100).build())
                .layer(2, new OutputLayer.Builder(LossFunctions.LossFunction.NEGATIVELOGLIKELIHOOD)
                        .activation(Activation.SOFTMAX).nIn(100).nOut(10).build())
                .pretrain(false).backprop(true)
                .build();

        //Configuration for Spark training: see http://deeplearning4j.org/spark for explanation of these configuration options
        TrainingMaster tm = new ParameterAveragingTrainingMaster.Builder(batchSizePerWorker)    //Each DataSet object: contains (by default) 32 examples
                .averagingFrequency(5)
                .workerPrefetchNumBatches(0)            //Async prefetching: 2 examples per worker
                .batchSizePerWorker(batchSizePerWorker)
                .storageLevel(StorageLevel.MEMORY_AND_DISK_SER())

                .build();

        //Create the Spark network
        SparkDl4jMultiLayer sparkNet = new SparkDl4jMultiLayer(sc, conf, tm);

        //Execute training:
        for (int i = 0; i < numEpochs; i++) {
            sparkNet.fit(trainData);
            log.info("Completed Epoch {}", i);
        }

        //Perform evaluation (distributed)
        Evaluation evaluation = sparkNet.evaluate(testData);
        log.info("***** Evaluation *****");
        log.info(evaluation.stats());

        //Delete the temp training files, now that we are done with them
        tm.deleteTempFiles(sc);

        log.info("***** Example Complete *****");
```

ERROR: 
```
17/07/03 22:37:08 ERROR TaskResultGetter: Exception while getting task result
com.esotericsoftware.kryo.KryoException: java.lang.NullPointerException
Serialization trace:
matrix (org.deeplearning4j.eval.ConfusionMatrix)
confusion (org.deeplearning4j.eval.Evaluation)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:144)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:551)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:708)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
        at com.esotericsoftware.kryo.serializers.FieldSerializer.read(FieldSerializer.java:551)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:790)
        at com.twitter.chill.SomeSerializer.read(SomeSerializer.scala:25)
        at com.twitter.chill.SomeSerializer.read(SomeSerializer.scala:19)
        at com.esotericsoftware.kryo.Kryo.readClassAndObject(Kryo.java:790)
        at org.apache.spark.serializer.KryoSerializerInstance.deserialize(KryoSerializer.scala:327)
        at org.apache.spark.scheduler.DirectTaskResult.value(TaskResult.scala:88)
        at org.apache.spark.scheduler.TaskResultGetter$$anon$3$$anonfun$run$1.apply$mcV$sp(TaskResultGetter.scala:72)
        at org.apache.spark.scheduler.TaskResultGetter$$anon$3$$anonfun$run$1.apply(TaskResultGetter.scala:63)
        at org.apache.spark.scheduler.TaskResultGetter$$anon$3$$anonfun$run$1.apply(TaskResultGetter.scala:63)
        at org.apache.spark.util.Utils$.logUncaughtExceptions(Utils.scala:1951)
        at org.apache.spark.scheduler.TaskResultGetter$$anon$3.run(TaskResultGetter.scala:62)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
        at java.util.Collections$SynchronizedMap.put(Collections.java:2588)
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:162)
        at com.esotericsoftware.kryo.serializers.MapSerializer.read(MapSerializer.java:39)
        at com.esotericsoftware.kryo.Kryo.readObject(Kryo.java:708)
        at com.esotericsoftware.kryo.serializers.ObjectField.read(ObjectField.java:125)
        ... 18 more
```

**Comment by jasstionzyf** [comment]
@huitseeker  how this issues going?

**Comment by AlexDBlack** [comment]
@jasstionzyf I'll take a look at this today

**Comment by jasstionzyf** [comment]
thanks very much, it seems spark kryo shaded jar has confilicts with nd4j-kryo_2.11 (0.8.1-SNAPSHOT)

**Comment by AlexDBlack** [comment]
@jasstionzyf Could I also have your full pom.xml file to help reproduce/debug this?

**Comment by jasstionzyf** [comment]
@AlexDBlack  ok, anyway https://gist.github.com/jasstionzyf/9f0784c04c1a579182dcd144abb2da4a   this is my all pom,  it all use 0.8.1-snapshot

**Comment by AlexDBlack** [comment]
Turns out: This is just Kryo failing on synchronized maps: this (with kryo configured) is enough to reproduce it...
```
        Map<Object,Object> m = Collections.synchronizedMap(new HashMap<>());
        m.put("Key", "Value");
        SerializerInstance si = sc.env().serializer().newInstance();
        ByteBuffer bb = si.serialize(m, null);
        Map<Object,Object> m2 = si.deserialize(bb, null);
```

The (well, *one possible*) fix is easy enough - don't use synchronized maps - but I'll do a more thorough check on other classes that Kryo would serialize during the course of training.

**Comment by huitseeker** [comment]
@AlexDBlack Could we just reuse https://github.com/magro/kryo-serializers ?

**Comment by AlexDBlack** [comment]
@huitseeker fyi: https://github.com/deeplearning4j/nd4j/pull/1913/files

@jasstionzyf Fix is tested + merged to master

@sethah I've been unable to reproduce your specific issue - I suspect it may depend on the specific type of map you are using for the learning rate schedule (but that isn't provided in the issue). Can you provide that - or at least enough that I can run/reproduce it?
Alternatively, try testing on snapshots - it's possible I've caught the issue with some of the kryo changes I've made.

**Comment by AlexDBlack** [comment]
Closing this due to inactivity + being unable to reproduce.
@sethah If you still have a problem, we can re-open this and I can look into it more.
